src/utils/useTodos.ts

```diff
@@ -0,0 +1,238 @@
+    const trimmed = title.trim();
+
+    if (!trimmed) {
+      showError('empty');
+      setErrorMessage(ERROR_MESSAGES.empty);
+      setIsAdding(false);
+
+      return;
+    }
+
+    const TEMP_ID = 0;
+    const temp: Todo = {
+      id: TEMP_ID,
+      title: trimmed,
+      completed: false,
+      userId: USER_ID,
+      isTemp: true,
+      todos: [],
+    };
+
+    setTempTodo(temp);
+    setSelectedTodoId(TEMP_ID);
+    setIsAdding(true);
+    setErrorMessage(null);
+
+    addTodo({ title: trimmed, completed: false, todos: [] })
+      .then(newTodo => {
+        setTodos(prev => {
+          const filtered = prev.filter(todo => todo.id !== TEMP_ID);
+
+          return [...filtered, newTodo];
+        });
+
+        setTempTodo(null);
+        setSelectedTodoId(null);
+        onSuccess?.();
+        setIsAdding(false);
+      })
+      .catch(() => {
+        showError('add');
+        setTimeout(() => {
+          setTempTodo(null);
+          setIsAdding(false);
+          setSelectedTodoId(null);
+        }, 4000);
+      });
+  };
+
+  const handleDelete = (id: number) => {
+    setDeletingTodoId(id);
+
+    deleteTodo(id)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== id));
+      })
+      .catch(() => {
+        showError('delete');
+      })
+      .finally(() => {
+        setTimeout(() => {
+          setDeletingTodoId(null);
+        }, 1000);
+      });
+  };
+
+  const handleToggleStatus = (id: number) => {
+    setTodos(prev =>
+      prev.map(todo =>
+        todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+      ),
+    );
+  };
+
+  const handleFormSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+    handleAddTodo(query, () => setQuery(''));
+  };
+
+  const onToggleAll = () => {
+    const allCompleted = todos.every(todo => todo.completed);
+
+    setIsLoading(true);
+
+    toggleAllTodos(todos, !allCompleted)
+      .then(() => {
+        setTodos(prevTodos =>
+          prevTodos.map(todo => ({ ...todo, completed: !allCompleted })),
+        );
+      })
+      .catch(() => {
+        showError('update');
+      })
+      .finally(() => {
+        setIsLoading(false);
+      });
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    try {
+      const results = await Promise.allSettled(
+        completedTodos.map(todo => deleteTodo(todo.id)),
+      );
+
+      const successfullyDeletedIds = completedTodos
+        .filter((_, index) => results[index].status === 'fulfilled')
+        .map(todo => todo.id);
+
+      setTodos(prev =>
+        prev.filter(todo => !successfullyDeletedIds.includes(todo.id)),
+      );
+
+      const hasErrors = results.some(r => r.status === 'rejected');
+
+      if (hasErrors) {
+        showError('delete');
+      }
+    } catch {
+      showError('delete');
+    }
+  };
+
+  const onFocusInput = () => {};
```

Comment:
Do you need this function?
Probably it should focus(`inputRef.current?.focus();`)?